cpp/src/arrow/array/array_test.cc

```diff
@@ -499,6 +501,24 @@ TEST_F(TestArray, TestMakeArrayOfNull) {
       }
     }
   }
+
+  for (int64_t length : {0, 16}) {
+    ARROW_SCOPED_TRACE("length = ", length, " (required fields)");
+
+    auto req = [](auto type) { return field("", std::move(type), /*nullable=*/false); };
+
+    // union with no nullable fields cannot represent a null
+    ASSERT_RAISES(Invalid, MakeArrayOfNull(dense_union({req(int8())}), length));
```

Comment:
Shouldn't this be successful if length is 0?

Reply:
If the type cannot support any number of nulls, I would say that the special case of zero length is not worth allowing